### ISSUE TITLE
Provide a mock vscode object for testing

### ISSUE THREAD

**Issue by Tbhesswebber** [microsoft/vscode#73640]
I am just now creating my first extension and the testing process has been horrendous and encourages such fragile testing that it's almost not worthwhile to test.

It would be great to see a test-double of the `vscode` object that contains some helper API's for testing in order to have cleaner, more stable tests and to make engineers more confident in the extensions that we write.

**Comment by jrieken** [comment]
> I am just now creating my first extension and the testing process has been horrendous and encourages such fragile testing that it's almost not worthwhile to test.

Not sure what we mean by this but we cannot do a second implementation of vscode for testing, it would likely differ too much from the real environment.

**Comment by Tbhesswebber** [comment]
That's one of the issues that I am running into with testing an extension - I need to run it within the context of multiple "environments" in order to make sure that the behavior is working as expected.  Testing the extension within the real environment is great _if_ you are running some kind of user acceptance test.

For example, a plugin that accepts a `myExtension.filepath` setting needs to be able to test the default file-path and one "provided by the user".  I have managed to get around this by using `fs` to change the test workspace on the fly, but it has resulted in hacky and brittle tests. This makes me lack confidence in the tests and, therefore, the code supported by them as I iterate on the product.

I can potentially mock out the specific methods that are being used under the hood by my extension. Even with that, in order for me to be confident that I'm not breaking something internally, I would need to inject the vscode object into the top level of my extension and use dependency injection all the way through so that I can inject a fully isolated test-double to each unit as I run my unit and integration tests.  This is where my feature request comes in: whether you are using dependency injection or mocking dependencies, a VSCode-supported mock of the vscode object that houses the API's and extends some test-only functionality would make it so that people can test their extensions with confidence and relative ease.

_Note: Maybe there's some testing pattern that is out there currently and I've missed it, but I looked through dozens of extensions to try to find healthy testing patterns and came up empty._

**Comment by jrieken** [comment]
The existing pattern is to the extension testing framework that we offer - that's the only environment of VS Code that exists. If you wanna test parts of your code without that then you need to abstract aways from the API defines in vscode.d.ts. That's comparable to browsers and library code that doesn't have browser dependencies. The latter you can (unit) test in any JS environment, the former requires a browser

**Comment by Tbhesswebber** [comment]
Code written for a browser does not require a browser, it requires an implementation of the browser API.  Unit and integration tests should be run in isolation, which requires a new instance of the webpage for every test or using JSDOM to allow tests to run in Node.  My feature request is to have a JSDOM (or Electron Spectron or Enzyme or ...) equivalent for VSCode so that we as engineers can have confidence in the tools we use and build every day.  These types of libraries that provide feature-rich environment equivalents specifically for testing are commonplace.

One of the reasons that I chose VS Code over other editors is because of the rich community and vast collection of extensions.  Now that I've discovered that so many extensions are untested, including ones created by Microsoft itself, I feel like I need to provide feedback about ways that I see to improve the testing experience in order to contribute to that rich community.
